app/middleware/decrypt_card_data.js

```diff
@@ -0,0 +1,65 @@
+'use strict'
+
+const logger = require('../utils/logger')(__filename)
+const decryptNode = require('@aws-crypto/decrypt-node')
+const rawRsaKeyringNode = require('@aws-crypto/raw-rsa-keyring-node')
+const cardDataFields = require('../../config/card_data_fields')
+
+function getMandatoryEnvironmentVariable (env, key) {
+  const result = env[key]
+  if (!result) {
+    throw new Error(`Environment variable ${key} is required when DECRYPT_AND_OMIT_CARD_DATA is set to true`)
+  }
+  return result
+}
+
+/**
+ * Creates a middleware function which will decrypt request parameters
+ * encrypted with AWS CloudFront's Field Level Encryption found in the request.
+ *
+ * Modifies req.body so handlers can deal with plaintext parameters.
+ *
+ * @param {Object} env - A map of environment variables.
+ *   DECRYPT_AND_OMIT_CARD_DATA      - if true, all other parameters are required. Otherwise returns an empty piece of middleware.
+ *   DECRYPT_CARD_DATA_PRIVATE_KEY   - a PEM formatted RSA private key
+ *   DECRYPT_CARD_DATA_KEY_NAME      - can be found on https://console.aws.amazon.com/cloudfront/home#publickey:
+ *   DECRYPT_CARD_DATA_KEY_NAMESPACE - the "Provider name" (NOT the "Profile name") from the FLE profile
+ *                                     https://console.aws.amazon.com/cloudfront/home#fle: shows a list of FLE profiles.
+ */
+function createMiddleware (env) {
+  const isEnabled = env.DECRYPT_AND_OMIT_CARD_DATA
+  if (isEnabled !== 'true') {
+    return function noopMiddelware (_req, _res, next) { next() }
+  }
+
+  const privateKey = getMandatoryEnvironmentVariable(env, 'DECRYPT_CARD_DATA_PRIVATE_KEY')
+  const keyName = getMandatoryEnvironmentVariable(env, 'DECRYPT_CARD_DATA_KEY_NAME')
+  const keyNamespace = getMandatoryEnvironmentVariable(env, 'DECRYPT_CARD_DATA_KEY_NAMESPACE')
+
+  const keyring = new rawRsaKeyringNode.RawRsaKeyringNode({
+    keyName,
+    keyNamespace,
+    rsaKey: { privateKey },
+    oaepHash: 'sha256'
+  })
+
+  return function decryptCardDataMiddleware (req, _res, next) {
+    const fieldsToDecrypt = cardDataFields.filter(x => req.body[x] !== undefined)
+    Promise.all(fieldsToDecrypt.map(async (fieldName) => {
+      const cipherText = req.body[fieldName]
+      try {
+        const result = await decryptNode.decrypt(keyring, Buffer.from(cipherText, 'base64'))
+        req.body[fieldName] = result.plaintext.toString('utf8')
+      } catch (rejection) {
+        logger.error('Failed to decrypt', { fieldName })
+        throw rejection
+      }
+    })).then(() => {
+      next()
+    }).catch((err) => {
+      next(err)
```

Comment:
This is absolutely how one should handle an error when writing middleware in a stack like this, my understanding is that these apps have a very dubious understanding of what error handling is and therefore might not actually have anything at the end of the stack that picks this up nicely. 

In that case it would default to Expresses error handler which would send plaintext with hopefully the stack redacted in production mode. 

Having had a quick look it looks like other middleware handle errors inline and respond immediately through the response object. This is not a good pattern and makes a response super hard to reason about but this PR might need to factor in that it would have to do that or add something at the end that will respond with errors in a uniform way.

Reply:
My understanding of what you'd be expecting at the other end but not actually find in frontend https://github.com/alphagov/pay-toolbox/blob/master/src/web/errorHandler.ts#L64-L87

Reply:
You're correct. Here's a screenshot with it snowflaked to error 😞 

![image](https://user-images.githubusercontent.com/1696784/75258734-d4a28b80-57de-11ea-8767-15187650bdc4.png)

Reply:
This should be addressed in 8e1a1da - I've added a default error handler. I've tested this manually in the paas staging environment and now you get the nice:

```
We’re experiencing technical problems

No money has been taken from your account.
```

thing instead.